include/swift/AST/DiagnosticsSIL.def

```diff
@@ -320,6 +320,23 @@ ERROR(negative_fp_literal_overflow_unsigned, none,
   "negative literal '%0' cannot be converted to %select{|unsigned }2%1",
   (StringRef, Type, bool))
 
+// Overflow and underflow warnings for floating-point truncation
+WARNING(warning_float_trunc_overflow, none,
+  "overflow: '%0' becomes %select{|-}3inf during%select{| implicit}2 "
+  "conversion to %1", (StringRef, Type, bool, bool))
+
+WARNING(warning_float_trunc_underflow, none,
+  "precision loss due to tininess during%select{| implicit}2 "
+  "conversion of '%0' to %1", (StringRef, Type, bool, bool))
+
+WARNING(warning_float_trunc_hex_inexact, none,
+"'%0' loses precision during%select{| implicit}2 conversion to %1",
```

Comment:
> the warning needs to be silenceable more ergonomically

Can you please elaborate on what that more elegant "ergonomic" syntax for disabling the warning would be in your opinion? If you can provide some examples where you do not want to see the warning it would be very helpful. 

(There is also the technical difficulty of whether we can detect those "erogonomic" subtleties at the SIL diagnostic level. Note that the AST information that we recover at SIL diagnostics level is limited.  E.g. `let x: Float = 1.0p26` ,`let x = (1.0p26  as Float)`, `let x = Float(1.0p26 )` all look pretty much the same at the SIL level even if we get the AST information from the SIL instructions, because of the way AST nodes are associated with SIL instructions.  The future changes in Swift will be helpful and has the potential to  make the diagnostics better. But as of now I am trying to make the diagnostics as much clear as possible without relying on the future changes.) 

But, let me know what kind of warning disabling you would like, and I will see if it can be detected and done with reasonable overheads and implementation at the SIL level.

Reply:
> '%0' overflows to %select{|-}3inf during%select{| implicit}2 conversion to $1
> '%0' loses precision during%select{| implicit}2 conversion to %1
> '%0' overflows to %select{|-}1inf because its magnitude exceeds the limits of a float literal

I am fine with all three of these messages.  I will make those changes. 

> '%0' underflows to %select{|-}30.0 during%select{| implicit}2 conversion to $1

I am not very sure about this. As far as I understand, 'underflow' is said to happen the moment a value becomes subnormal i.e, the exponent in the normal form falls below minimum representable value. But the warning is emitted only when this underflow results in loss of precision _more than_ what would happen without an underflow. 
In other words, 3 conditions must hold for the warning: (a) degeneration to subnormal form (underflow) and (b)  loss of precision during conversion (inexactness) and (c) loss of precision more than what would happen with unbounded exponent. 
  (For more information see this discussion: https://forums.swift.org/t/rfc-on-compile-time-diagnosis-of-erroneous-floating-point-operations/12125/11).  

It has been difficult for me to come up with a succinct, good message here.  But, I feel that the current message provides more information to a user than just that an "underflow" happened.  It is meant to convey that there is more imprecision because of the underflow than what they would expect from the truncation.

I would think that something like the following would be better:
"'%0' loses precision due to tininess during%select{| implicit}2 conversion to $1"

Btw, although I would like to print the new value (in your example -30.0 ), it may not convey a loss in precision. E.g. try in REPL

(swift) let _: Float = 5.87747e-39
<REPL Input>:1:16: warning: precision loss due to tininess during conversion of '5.87747e-39' to 'Float'
let _: Float = 5.87747e-39
               ^
// _ : Float = 5.87747e-39

Here the new and the old values look identical (because of the way they are printed) but there has been a loss in precision. I am wondering how to print the value, perhaps we could go one enumerating digits until it starts differing from the literal written by the user. But not sure if it is worth the effort.

Reply:
@xwu  Btw, if you would like any changes to logic/conditions that are used to decide on when to warn on underflows, please post in the Swift forum discussion thread presented above. That way there are also other experts in the thread, who can be brought into the discussion.

Reply:
I don't want to answer for @xwu, but I think we need a way for a user to indicate *locally* that loss of precision is ok for a specific literal, without disabling the warning by a flag at file or module scope.

Reply:
@stephentyrone @xwu  Thanks for the feedback. I have pushed a new commit in which the warnings are disabled if the floating-point literal is truncated using an explicit initializer. 
E.g.

>  Float(5.87747e-39)  // no warning
> let f: Float = 5.87747e-39 
 warning: '5.87747e-39' underflows and loses precision during conversion to 'Float'